### PR TITLE
Syw UID2-3679 cpp sdk base64 v4 token

### PR DESCRIPTION
**Pull request by sunnywu**
1. Able to handle Base64 (non url friendly) encoded v4 ad token decryption and added a unit test
2. I use Clion/Docker/Mac OS setup so added some readme entries how people can replicate such setup easily

### CODE REVIEW

File: lib/uid2encryption.cpp [hunk 1]
```diff
@@ -45,7 +45,8 @@ DecryptionResult DecryptToken(const std::string& token, const KeyContainer& keys
     }
 
     const std::string headerStr = token.substr(0, 4);
-    const bool isBase64UrlEncoding = std::any_of(headerStr.begin(), headerStr.end(), [](char c) { return c == '-' || c == '_'; });
+    // check the whole ad token string instead of the headerStr to make sure
```
Comment (jon8787) [inline lib/uid2encryption.cpp:47]: nit - `headerStr` should be moved to where it's first used (above line 54)

File: lib/uid2encryption.cpp [hunk 2]
```diff
@@ -69,8 +70,13 @@ DecryptionResult DecryptToken(const std::string& token, const KeyContainer& keys
             return DecryptTokenV3(encryptedId, keys, now, identityScope, checkValidity);
         }
         if (headerBytes[1] == static_cast<std::uint8_t>(AdvertisingTokenVersion::V4)) {
-            // same as V3 but use Base64URL encoding
-            uid2::UID2Base64UrlCoder::Decode(token, encryptedId);
+            if (isBase64UrlEncoding) {
+                // same as V3 but use Base64URL encoding
+                uid2::UID2Base64UrlCoder::Decode(token, encryptedId);
```
Comment (jon8787) [inline lib/uid2encryption.cpp:75]: This is different from the [Java SDK](https://github.com/IABTechLab/uid2-client-java/blob/main/src/main/java/com/uid2/client/Uid2Encryption.java#L47) which converts the token to  to base64 and then calls decryptV3. 
Is that intentional?

File: test/encryption_tests_v4.cpp [hunk 3]
```diff
@@ -116,6 +117,38 @@ std::string GenerateUid2TokenV4AndValidate(
     return advertisingToken;
 }
 
+TEST(EncryptionTestsV4, CanDecryptV4TokenEncodedAsBase64)
+{
+    UID2Client client("ep", "ak", CLIENT_SECRET, IdentityScope::UID2);
+    client.RefreshJson(KeySetToJson({MASTER_KEY, SITE_KEY}));
+    auto advertisingToken = GenerateUid2TokenV4AndValidate(EXAMPLE_UID, MASTER_KEY, SITE_ID, SITE_KEY, EncryptTokenParams());
```
Comment (jon8787) [inline test/encryption_tests_v4.cpp]: can we use do/while so we only have 1 `GenerateUid2TokenV4AndValidate` like in [C# SDK](https://github.com/IABTechLab/uid2-client-net/pull/62/files#diff-b0efbb43a718e133eeb1f9fc235aa4c74f3cb439161d8ccc64041a11802bb135R94)?

File: test/encryption_tests_v4.cpp [hunk 4]
```diff
@@ -116,6 +117,38 @@ std::string GenerateUid2TokenV4AndValidate(
...
+            break;
+        }
+        advertisingToken = GenerateUid2TokenV4AndValidate(EXAMPLE_UID, MASTER_KEY, SITE_ID, SITE_KEY, EncryptTokenParams());
+    }
+
+    const bool isBase64UrlEncoding = std::any_of(advertisingToken.begin(), advertisingToken.end(), [](char c) { return c == '-' || c == '_'; });
+    EXPECT_TRUE(isBase64UrlEncoding);
+
+    std::vector<std::uint8_t> adTokenBytes;
+    uid2::UID2Base64UrlCoder::Decode(advertisingToken, adTokenBytes);
```
Comment (jon8787) [inline test/encryption_tests_v4.cpp:140]: Is any of the below code duplicated from the other tests?
Worth making a function?

File: lib/uid2encryption.cpp (hunk 2 above)
Comment (sunnywu) [reply lib/uid2encryption.cpp:75]: not intentional but i just chose the easiest way - don't necessarily see the need to clone the exact same logic as long as we have unit tests to cover it?

File: test/encryption_tests_v4.cpp (hunk 3 above)
Comment (sunnywu) [reply test/encryption_tests_v4.cpp]: no i tried that initially

Comment (jon8787) [reply test/encryption_tests_v4.cpp]: what was the issue / error?

File: lib/uid2encryption.cpp (hunk 1 above)
Comment (sunnywu) [reply lib/uid2encryption.cpp:47]: fixed

File: test/encryption_tests_v4.cpp (hunk 4 above)
Comment (sunnywu) [reply test/encryption_tests_v4.cpp:140]: done some refactoring to have DecryptAndAssertSuccess method like other SDK

File: test/encryption_tests_v4.cpp (hunk 3 above)
Comment (sunnywu) [reply test/encryption_tests_v4.cpp]: hmm i changed to do the base64 char check inline within while condition and now it works

File: test/encryption_tests_v4.cpp [hunk 5]
```diff
@@ -116,15 +117,46 @@ std::string GenerateUid2TokenV4AndValidate(
     return advertisingToken;
 }
 
+void DecryptAndAssertSuccess(UID2Client& client, const std::string& advertisingTokenString, Timestamp timestamp = Timestamp::Now())
```
Comment (jon8787) [inline test/encryption_tests_v4.cpp:120]: does `const UID2Client&` work?

File: test/encryption_tests_v4.cpp [hunk 6]
```diff
@@ -116,15 +117,46 @@ std::string GenerateUid2TokenV4AndValidate(
...
+    UID2Client client("ep", "ak", CLIENT_SECRET, IdentityScope::UID2);
+    client.RefreshJson(KeySetToJson({MASTER_KEY, SITE_KEY}));
+    std::string advertisingToken;
+
+    // for testing purposes, the token must have some Base64URL encoding characters
+    do {
+        advertisingToken = GenerateUid2TokenV4AndValidate(EXAMPLE_UID, MASTER_KEY, SITE_ID, SITE_KEY, EncryptTokenParams());
+    } while (!std::any_of(advertisingToken.begin(), advertisingToken.end(), [](char c) { return c == '-' || c == '_'; }));
+
+    const bool isBase64UrlEncoding = std::any_of(advertisingToken.begin(), advertisingToken.end(), [](char c) { return c == '-' || c == '_'; });
```
Comment (jon8787) [inline test/encryption_tests_v4.cpp]: I think we don't need this anymore because of the while condition

**Review by jon8787 — APPROVED** [review]

File: test/encryption_tests_v4.cpp (hunk 5 above)
Comment (sunnywu) [reply test/encryption_tests_v4.cpp:120]: tried but no

**Pull request merged**